cpp/oneapi/dal/table/detail/csr.hpp

```diff
@@ -0,0 +1,255 @@
+    /// An empty table (see the default constructor) has a unique ``kind`` value as well.
+    /// @remark default = empty_table_kind
+    std::int64_t get_kind() const;
+
+    /// The layout of the data within the table
+    /// @remark default = data_layout::csr
+    data_layout get_data_layout() const;
+
+protected:
+    sparse_table(pimpl::element_type* impl);
+    void sparse_init_impl(pimpl::element_type* impl);
+
+private:
+    pimpl impl_;
+};
+
+class ONEDAL_EXPORT csr_table : public sparse_table {
+    friend detail::pimpl_accessor;
+    using pimpl = detail::pimpl<dal::backend::csr_table_impl>;
+
+public:
+    /// Creates a new ``csr_table`` instance with zero number of rows and columns.
+    /// The :expr:`kind` is set to``csr_table::kind()``.
+    /// All the properties should be set to default values (see the Properties section).
+    csr_table();
+
+    /// Creates a new ``csr_table`` instance from externally-defined data blocks.
+    /// Table object owns the data, row indices and column indices pointers.
+    ///
+    /// @tparam Data          The type of elements in the data block that will be stored into the table.
+    ///                       The :literal:`Data` type should be at least :expr:`float`, :expr:`double` or :expr:`std::int32_t`.
+    /// @tparam DataDeleter   The type of a deleter called on ``data`` when
+    ///                       the last table that refers it is out of the scope.
+    /// @tparam ColumnDeleter The type of a deleter called on ``column_indices`` when
+    ///                       the last table that refers it is out of the scope.
+    /// @tparam RowDeleter    The type of a deleter called on ``row_indices`` when
+    ///                       the last table that refers it is out of the scope.
+    ///
+    /// @param data           The pointer to values in the CSR layout.
+    /// @param column_indices The pointer to column indices in the CSR layout.
+    /// @param row_indices    The pointer to row indices in the CSR layout.
+    /// @param column_count   The number of columns in the corresponding dense table.
+    /// @param row_count      The number of rows in the corresponding dense table.
+    /// @param data_deleter   The deleter that is called on the ``data`` when the last table that refers it
+    ///                       is out of the scope.
+    /// @param column_deleter The deleter that is called on the ``column_indices`` when the last table that refers it
+    ///                       is out of the scope.
+    /// @param row_deleter    The deleter that is called on the ``row_indices`` when the last table that refers it
+    ///                       is out of the scope.
+    /// @param indexing       The indexing scheme used to access data in the CSR layout. Support only :literal:`csr_indexing::one_based`.
+    template <typename Data, typename DataDeleter, typename ColumnDeleter, typename RowDeleter>
+    csr_table(const Data* data,
```

Comment:
Can we replace this ctor with ctor accepting `dal::array`s?

Reply:
The comment is still valid

Reply:
Why you need to align it with homogen table? It's much simpler to have a ctor from arrays